% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/limma_contrasts.R
\name{limma_contrasts}
\alias{limma_contrasts}
\title{Test Custom Contrasts with LIMMA}
\usage{
limma_contrasts(
  eset,
  model.str,
  coef.str,
  contrasts,
  stat = c("t", "F"),
  var.group = character(0),
  plot = FALSE,
  col.group = coef.str,
  adjust.method = "BH",
  remove_name = TRUE,
  ...
)
}
\arguments{
\item{eset}{\code{eSet} (or most likely \code{eSet} subclass) object}

\item{model.str}{character; formulation of the model (e.g. \code{"~ 0 + a +
  b"}). This should be a no-intercept model (it should include 0 or -1 as
terms). See \code{\link[stats]{lm}} for more details.}

\item{coef.str}{character; coefficient of interest. One of
\code{colnames(pData(eset))}. E.g. \code{"a"}.}

\item{contrasts}{character; contrasts to test. All factor levels in the
contrasts should begin with the name of that factor. It is recommended to
use \code{\link{paircomp}} with the \code{name} argument specified to
create pairwise contrasts of the form \code{"a-b"}.}

\item{stat}{character; the moderated test statistic to use. Either \code{"t"}
or \code{"F"}.}

\item{var.group}{\code{NULL} or character; the column in \code{pData(eset)}
indicating groups that will have different relative quality weights. The
default (\code{NULL}), means that each sample will be weighted equally.}

\item{plot}{logical; whether to generate diagnostic plots. The default
(\code{TRUE}) generates a barplot of weights applied to each sample and a
plot of the mean-variance trend using \code{\link[limma]{plotSA}}.}

\item{col.group}{character; the column in \code{pData(eset)} used to color
the text of the MDS plot.}

\item{adjust.method}{method for p-value adjustment. Default is \code{"BH"}
(Benjamini-Hochberg). See \code{\link[stats]{p.adjust}} for details.}

\item{remove_name}{if \code{TRUE} (the default) \code{coef.str} will be
removed from the contrasts. This simply improves readability, but it will
not work properly if \code{coef.str} is a substring of any group name.}

\item{...}{additional arguments passed to \code{\link[limma]{eBayes}}.}
}
\value{
\code{data.frame}. Basically the output of
\code{\link[limma]{topTable}} with additional columns \code{feature} and
\code{contrast} (if \code{stat = "t"}). All columns from \code{fData} are
also included.
}
\description{
A convenience wrapper for differential analysis with LIMMA using custom
contrasts.
}
\details{
An MDS plot is used to determine the appropriate value of
\code{var.group}. If samples within phenotype groups tend to cluster well
and have similar dispersions, the default \code{var.group = NULL} is
recommended. If samples within phenotype groups tend to cluster well, but
different groups are more or less dispersed, it may be beneficial to set
\code{var.group} to the name of the phenotype column. If one or more
samples tend to cluster poorly with samples of the same phenotype, it may
be beneficial to weight by sample. That is, set \code{var.group} to the
name of the column in \code{pData(eset)} that uniquely identifies each
sample. If variation between samples or groups of samples is biological
rather than technical, weighting is not recommended.

The plot of the mean-variance trend helps determine whether to fit a trend
line to the prior variance (default is \code{trend = FALSE}). It also helps
determine if the prior variance should be robustified against outlier
sample variances (\code{robust = TRUE}; default is \code{FALSE}). See
\code{\link[limma]{eBayes}} for more details.

p-values are adjusted across all contrasts (globally). It is recommended to
adjust p-values from similar contrasts together. See the "Multiple Testing
Across Contrasts" section of the LIMMA User's Guide (linked below) for more
information.
}
\examples{
# library(MSnSet.utils)
data(cptac_oca)

# A no-intercept model is required. Testing SUBTYPE contrasts
# with AGE as covariate
model.str <- "~ 0 + SUBTYPE + AGE"
coef.str <- "SUBTYPE"

# Create contrasts
contrasts <- paircomp(pData(oca.set)[, coef.str], name = coef.str)

# Moderated t-statistics
res1 <- limma_contrasts(oca.set, model.str = model.str,
                        coef.str = coef.str,
                        contrasts = contrasts)

# Moderated F-statistics
res2 <- limma_contrasts(oca.set, model.str = model.str,
                        coef.str = coef.str,
                        contrasts = contrasts, stat = "F")

# Count number of significant (< 0.05) features
sum(res2$adj.P.Val < 0.05)

# Robustify mean-variance trend (based on MDS plot)
res3 <- limma_contrasts(oca.set, model.str = model.str,
                        coef.str = coef.str,
                        contrasts = contrasts,
                        trend = TRUE, robust = TRUE)

# Count number of significant (< 0.05) features in each contrast
table(res3$contrast, res3$adj.P.Val < 0.05)

}
\references{
Liu, R., Holik, A. Z., Su, S., Jansz, N., Chen, K., Leong, H. S.,
Blewitt, M. E., Asselin-Labat, M. L., Smyth, G. K., & Ritchie, M. E.
(2015). Why weight? Modelling sample and observational level variability
improves power in RNA-seq analyses. \emph{Nucleic acids research}, \emph{43}(15),
e97. \url{https://doi.org/10.1093/nar/gkv412}

Phipson, B., Lee, S., Majewski, I. J., Alexander, W. S., & Smyth, G. K.
(2016). ROBUST HYPERPARAMETER ESTIMATION PROTECTS AGAINST HYPERVARIABLE
GENES AND IMPROVES POWER TO DETECT DIFFERENTIAL EXPRESSION. \emph{The annals of
applied statistics}, \emph{10}(2), 946--963.
\url{https://doi.org/10.1214/16-AOAS920}

Smyth G. K. (2004). Linear models and empirical bayes methods for assessing
differential expression in microarray experiments. \emph{Statistical
applications in genetics and molecular biology}, \emph{3}, Article3.
\url{https://doi.org/10.2202/1544-6115.1027}
}
\seealso{
\href{https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf}{LIMMA
User's Guide}, \code{\link{limma_a_b}}, \code{\link{limma_gen}}
}
